kratos/solving_strategies/schemes/residual_based_bdf_scheme.h

```diff
@@ -0,0 +1,582 @@
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+    
+    /**
+     * @brief Performing the update of the derivatives
+     * @param rModelPart The model of the problem to solve
+     * @param rDofSet Set of all primary variables
+     * @param A LHS matrix
+     * @param Dx incremental update of primary variables
+     * @param b RHS Vector
+     */
+    inline void UpdateDerivatives(
+        ModelPart& rModelPart,
+        DofsArrayType& rDofSet,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b
+        )
+    {
+        // Updating time derivatives (nodally for efficiency)
+        const int num_nodes = static_cast<int>( rModelPart.Nodes().size() );
+
+        #pragma omp parallel for
+        for(int i = 0;  i< num_nodes; ++i) {
+            auto it_node = rModelPart.Nodes().begin() + i;
+            
+            UpdateFirstDerivative(it_node);
+            UpdateSecondDerivative(it_node);
+        }
+    }
+    
+    /**
+     * @brief Updating first time derivative (velocity)
+     * @param itNode the node interator
+     */
+    
+    virtual inline void UpdateFirstDerivative(NodesArrayType::iterator itNode)
+    {
+        KRATOS_ERROR << "Calling base BDF class" << std::endl;
+    }
+
+    /**
+     * @brief Updating second time derivative (acceleration)
+     * @param itNode the node interator
+     */
+    
+    virtual inline void UpdateSecondDerivative(NodesArrayType::iterator itNode)
+    {
+        KRATOS_ERROR << "Calling base BDF class" << std::endl;
+    }
+    
+    /**
+     * @brief It adds the dynamic LHS contribution of the elements 
+     * \f[ LHS = \frac{d(-RHS)}{d(u_{n0})} = c_0^2\mathbf{M} + c_0 \mathbf{D} + \mathbf{K} \f]
+     * @param LHS_Contribution The dynamic contribution for the LHS
+     * @param D The damping matrix
+     * @param M The mass matrix
+     * @param rCurrentProcessInfo The current process info instance
+     */
+
+    void AddDynamicsToLHS(
+        LocalSystemMatrixType& LHS_Contribution,
+        LocalSystemMatrixType& D,
+        LocalSystemMatrixType& M,
+        ProcessInfo& rCurrentProcessInfo
+        ) override
+    {
+        // Adding mass contribution to the dynamic stiffness
+        if (M.size1() != 0) { // if M matrix declared
+            noalias(LHS_Contribution) += M * std::pow(mBDF[0], 2);
+        }
+
+        // Adding  damping contribution
+        if (D.size1() != 0) { // if D matrix declared
+            noalias(LHS_Contribution) += D * mBDF[0];
+        }
+    }
+
+    /**
+     * @brief It adds the dynamic RHS contribution of the elements 
+     * \f[ \mathbf{b} - \mathbf{M} a - \mathbf{D} v \f]
+     * @param pElement The element to compute
+     * @param RHS_Contribution The dynamic contribution for the RHS
+     * @param D The damping matrix
+     * @param M The mass matrix
+     * @param rCurrentProcessInfo The current process info instance
+     */
+
+    void AddDynamicsToRHS(
```

Comment:
why don't you create a templated function that is called inside the "AddDynamicsToRHS" which takes the element or condition as template => this would reduce code duplication

Reply:
Good idea

Reply:
It is complicated to create a template, because you cannot combine virtual with template for methods (I thin I found a solution but doesn't reduce the code that much and increases the complexity)

Reply:
No, it is not possible because of the virtuals+templates

Reply:
I know
What I meant is the following:
~~~cpp
virtual AddDynamicsToRHS(Element)
{
    TemplatedAddDynamicsToRHS(Element);
}
virtual AddDynamicsToRHS(Condition)
{
    TemplatedAddDynamicsToRHS(Condition);
}

template <typename TObjectType>
TemplatedAddDynamicsToRHS(TObjectType rObject) // NOT virtual
{
    Do the Stuff
}
~~~

Sth like this

Reply:
I tried that, the problem is that method is calling derived methods

Reply:
Sorry I tried a different thing, trying now

Reply:
I did it, see [here](https://github.com/KratosMultiphysics/Kratos/blob/matrix-based-mapping-dev/applications/MappingApplication/custom_strategies/builders/ublas_mapping_matrix_builder.h#L122)
My template is the variable

Reply:
Yes, done